test-data/unit/check-newsemanal.test

```diff
@@ -1580,6 +1580,40 @@ class C:
 def C() -> None:  # E: Name 'C' already defined on line 1
     ''()  # E: "str" not callable
 
+[case testNewAnalyzerTupleIteration]
+from typing import Union, Tuple, NamedTuple
+
+class A: pass
+class B(A): pass
+class C(A): pass
+
+class T(Tuple[B, C]):
+    pass
+
+class NTInt(NamedTuple):
+    x: int
+    y: int
+
+class NTStr(NamedTuple):
+    x: str
+    y: str
```

Comment:
I would add couple non-trivial upper bound checks, just be sure they behave reasonably with the current ordering (fallbacks first, then bounds).